File: Windows_and_Linux/codex_auth.py
"""
Codex OAuth Authentication
--------------------------

Handles PKCE generation, authorization URL construction, token exchange,
token refresh, and JWT parsing for the OpenAI Codex OAuth flow.
"""

import base64
import hashlib
import json
import logging
import os
import secrets
import string
import time
from dataclasses import dataclass
from typing import Optional
from urllib.parse import urlencode

import requests

# ── Constants ────────────────────────────────────────────────────

CLIENT_ID = "app_EMoamEEZ73f0CkXaXp7hrann"
ISSUER = "https://auth.openai.com"
REDIRECT_PORT = 1455
REDIRECT_URI = f"http://localhost:{REDIRECT_PORT}/auth/callback"
SCOPES = "openid profile email offline_access"
RESPONSES_ENDPOINT = "https://chatgpt.com/backend-api/codex/responses"
CALLBACK_TIMEOUT_SECONDS = 5 * 60  # 5 minutes


# ── PKCE ─────────────────────────────────────────────────────────

@dataclass
class PkceCodes:
    verifier: str
    challenge: str


def generate_pkce() -> PkceCodes:
    """Generate PKCE code verifier (43 chars, RFC 7636) and S256 challenge."""
    chars = string.ascii_letters + string.digits + "-._~"
    verifier = "".join(secrets.choice(chars) for _ in range(43))
    digest = hashlib.sha256(verifier.encode("utf-8")).digest()
    challenge = base64.urlsafe_b64encode(digest).rstrip(b"=").decode("ascii")
    return PkceCodes(verifier=verifier, challenge=challenge)


def generate_state() -> str:
    """Generate a random OAuth state parameter (32 bytes, base64url)."""
    return base64.urlsafe_b64encode(secrets.token_bytes(32)).rstrip(b"=").decode("ascii")


# ── Authorization URL ────────────────────────────────────────────

def build_authorize_url(pkce: PkceCodes, state: str) -> str:
    params = {
        "response_type": "code",
        "client_id": CLIENT_ID,
        "redirect_uri": REDIRECT_URI,
        "scope": SCOPES,
        "code_challenge": pkce.challenge,
        "code_challenge_method": "S256",
        "id_token_add_organizations": "true",
        "codex_cli_simplified_flow": "true",
        "state": state,
        "originator": "writing-tools-desktop",
    }
    return f"{ISSUER}/oauth/authorize?{urlencode(params)}"


# ── OAuth Tokens ─────────────────────────────────────────────────

@dataclass
class OAuthTokens:
    access_token: str
    refresh_token: str
    expires_at: float  # Unix timestamp
    account_id: Optional[str] = None

    @property
    def is_expired(self) -> bool:
        return time.time() >= self.expires_at

    @property
    def is_expiring_soon(self) -> bool:
        return time.time() + 300 >= self.expires_at  # within 5 minutes

    def to_dict(self) -> dict:
        return {
            "access_token": self.access_token,
            "refresh_token": self.refresh_token,
            "expires_at": self.expires_at,
            "account_id": self.account_id,
        }

    @staticmethod
    def from_dict(d: dict) -> "OAuthTokens":
        return OAuthTokens(
            access_token=d["access_token"],
            refresh_token=d["refresh_token"],
            expires_at=d.get("expires_at", 0.0),
            account_id=d.get("account_id"),
        )


# ── Token Exchange ───────────────────────────────────────────────

def exchange_code_for_tokens(code: str, pkce_verifier: str) -> OAuthTokens:
    """Exchange an authorization code for OAuth tokens."""
    resp = requests.post(
        f"{ISSUER}/oauth/token",
        data={
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": REDIRECT_URI,
            "client_id": CLIENT_ID,
            "code_verifier": pkce_verifier,
        },
        headers={"Content-Type": "application/x-www-form-urlencoded"},
        timeout=30,
    )

    if not resp.ok:
        logging.error(f"Token exchange failed: {resp.status_code} — {resp.text}")
        raise RuntimeError(f"Token exchange failed: HTTP {resp.status_code}")

    data = resp.json()
    expires_in = data.get("expires_in", 3600)
    account_id = _parse_account_id(data.get("id_token"))

    return OAuthTokens(
        access_token=data["access_token"],
        refresh_token=data["refresh_token"],
        expires_at=time.time() + expires_in,
        account_id=account_id,
    )


def refresh_access_token(refresh_token: str) -> OAuthTokens:
    """Refresh an expired access token."""
    resp = requests.post(
        f"{ISSUER}/oauth/token",
        data={
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
            "client_id": CLIENT_ID,
        },
        headers={"Content-Type": "application/x-www-form-urlencoded"},
        timeout=30,
    )

    if not resp.ok:
        logging.error(f"Token refresh failed: {resp.status_code} — {resp.text}")
        raise RuntimeError(f"Token refresh failed: HTTP {resp.status_code}")

    data = resp.json()
    expires_in = data.get("expires_in", 3600)
    account_id = _parse_account_id(data.get("id_token"))

    return OAuthTokens(
        access_token=data["access_token"],
        refresh_token=data["refresh_token"],
        expires_at=time.time() + expires_in,
        account_id=account_id,
    )


# ── JWT Parsing ──────────────────────────────────────────────────

def _parse_account_id(id_token: Optional[str]) -> Optional[str]:
    """Extract the account ID from a JWT id_token."""
    if not id_token:
        return None
    try:
        parts = id_token.split(".")
        if len(parts) != 3:
            return None
        # Add padding for base64
        payload_b64 = parts[1]
        padding = 4 - len(payload_b64) % 4
        if padding != 4:
            payload_b64 += "=" * padding
        payload = json.loads(base64.urlsafe_b64decode(payload_b64))

        # Try multiple sources
        if "chatgpt_account_id" in payload:
            return payload["chatgpt_account_id"]
        auth_claims = payload.get("https://api.openai.com/auth", {})
        if isinstance(auth_claims, dict) and "chatgpt_account_id" in auth_claims:
            return auth_claims["chatgpt_account_id"]
        orgs = payload.get("organizations")
        if isinstance(orgs, list) and orgs:
            return orgs[0].get("id")
        return payload.get("email")
    except Exception as e:
        logging.warning(f"Failed to parse JWT: {e}")
        return None

File: macOS/WritingTools/Models/CodexAuth.swift
//
//  CodexAuth.swift
//  WritingTools
//
//  OAuth utilities for OpenAI Codex authentication
//

import Foundation
import CryptoKit

private let logger = AppLogger.logger("CodexAuth")

enum CodexAuthError: LocalizedError {
    case invalidState
    case missingCode
    case browserLaunchFailed
    case tokenExchangeFailed(String)
    case tokenRefreshFailed(String)
    case invalidToken
    
    var errorDescription: String? {
        switch self {
        case .invalidState:
            return "OAuth state validation failed"
        case .missingCode:
            return "Authorization code not received"
        case .browserLaunchFailed:
            return "Couldn't open the browser to continue authorization"
        case .tokenExchangeFailed(let message):
            return "Token exchange failed: \(message)"
        case .tokenRefreshFailed(let message):
            return "Token refresh failed: \(message)"
        case .invalidToken:
            return "Invalid token format"
        }
    }
}

struct CodexPkceCodes {
    let verifier: String
    let challenge: String
}

class CodexAuth {
    
    // MARK: - PKCE Generation
    
    static func generatePkce() -> CodexPkceCodes {
        let verifier = generateRandomString(length: 43)
        let challenge = generateChallenge(from: verifier)
        return CodexPkceCodes(verifier: verifier, challenge: challenge)
    }
    
    private static func generateRandomString(length: Int) -> String {
        let chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"
        var random = SystemRandomNumberGenerator()
        return String((0..<length).map { _ in
            chars.randomElement(using: &random)!
        })
    }
    
    private static func generateChallenge(from verifier: String) -> String {
        guard let data = verifier.data(using: .utf8) else {
            fatalError("Failed to encode verifier")
        }
        let hash = SHA256.hash(data: data)
        return Data(hash).base64URLEncodedString()
    }
    
    // MARK: - State Generation
    
    static func generateState() -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes).base64URLEncodedString()
    }
    
    // MARK: - Authorization URL
    
    static func buildAuthorizeUrl(pkce: CodexPkceCodes, state: String) -> URL {
        var components = URLComponents(string: "\(CodexConstants.issuer)/oauth/authorize")!
        components.queryItems = [
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "client_id", value: CodexConstants.clientId),
            URLQueryItem(name: "redirect_uri", value: CodexConstants.redirectUri),
            URLQueryItem(name: "scope", value: CodexConstants.scopes),
            URLQueryItem(name: "code_challenge", value: pkce.challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "id_token_add_organizations", value: "true"),
            URLQueryItem(name: "codex_cli_simplified_flow", value: "true"),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "originator", value: "writing-tools-macos")
        ]
        return components.url!
    }
    
    // MARK: - Token Exchange
    
    static func exchangeCodeForTokens(code: String, pkceVerifier: String) async throws -> OAuthTokens {
        let url = URL(string: "\(CodexConstants.issuer)/oauth/token")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        
        let body = [
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": CodexConstants.redirectUri,
            "client_id": CodexConstants.clientId,
            "code_verifier": pkceVerifier
        ]
        
        request.httpBody = body.formURLEncoded()
        
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw CodexAuthError.tokenExchangeFailed("Invalid response")
            }
            
            guard (200...299).contains(httpResponse.statusCode) else {
                let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown error"
                logger.error("Token exchange failed: \(httpResponse.statusCode) - \(errorMessage)")
                throw CodexAuthError.tokenExchangeFailed("HTTP \(httpResponse.statusCode)")
            }
            
            let tokenResponse = try JSONDecoder().decode(CodexTokenResponse.self, from: data)
            
            // Extract account ID from ID token if available
            let accountId = tokenResponse.idToken.flatMap { parseJwtClaims(token: $0) }
            
            return OAuthTokens(
                accessToken: tokenResponse.accessToken,
                refreshToken: tokenResponse.refreshToken,
                expiresIn: tokenResponse.expiresIn ?? 3600,
                accountId: accountId
            )
            
        } catch let error as CodexAuthError {
            throw error
        } catch {
            logger.error("Token exchange error: \(error.localizedDescription)")
            throw CodexAuthError.tokenExchangeFailed(error.localizedDescription)
        }
    }
    
    // MARK: - Token Refresh
    
    static func refreshAccessToken(refreshToken: String) async throws -> OAuthTokens {
        let url = URL(string: "\(CodexConstants.issuer)/oauth/token")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        
        let body = [
            "grant_type": "refresh_token",
            "refresh_token": refreshToken,
            "client_id": CodexConstants.clientId
        ]
        
        request.httpBody = body.formURLEncoded()
        
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw CodexAuthError.tokenRefreshFailed("Invalid response")
            }
            
            guard (200...299).contains(httpResponse.statusCode) else {
                let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown error"
                logger.error("Token refresh failed: \(httpResponse.statusCode) - \(errorMessage)")
                throw CodexAuthError.tokenRefreshFailed("HTTP \(httpResponse.statusCode)")
            }
            
            let tokenResponse = try JSONDecoder().decode(CodexTokenResponse.self, from: data)
            
            // Extract account ID from ID token if available
            let accountId = tokenResponse.idToken.flatMap { parseJwtClaims(token: $0) }
            
            return OAuthTokens(
                accessToken: tokenResponse.accessToken,
                refreshToken: tokenResponse.refreshToken,
                expiresIn: tokenResponse.expiresIn ?? 3600,
                accountId: accountId
            )
            
        } catch let error as CodexAuthError {
            throw error
        } catch {
            logger.error("Token refresh error: \(error.localizedDescription)")
            throw CodexAuthError.tokenRefreshFailed(error.localizedDescription)
        }
    }
    
    // MARK: - JWT Parsing
    
    static func parseJwtClaims(token: String) -> String? {
        let parts = token.split(separator: ".")
        guard parts.count == 3 else {
            logger.warning("Invalid JWT format")
            return nil
        }
        
        guard let payloadData = Data(base64URLEncoded: String(parts[1])) else {
            logger.warning("Failed to decode JWT payload")
            return nil
        }
        
        do {
            let claims = try JSONDecoder().decode(CodexIdTokenClaims.self, from: payloadData)
            return extractAccountId(from: claims)
        } catch {
            logger.warning("Failed to parse JWT claims: \(error.localizedDescription)")
            return nil
        }
    }
    
    private static func extractAccountId(from claims: CodexIdTokenClaims) -> String? {
        // Try multiple sources for account ID
        if let accountId = claims.chatgptAccountId {
            return accountId
        }
        if let accountId = claims.customClaims?.chatgptAccountId {
            return accountId
        }
        if let orgId = claims.organizations?.first?.id {
            return orgId
        }
        return claims.email
    }
}

// MARK: - Helper Extensions

extension Dictionary where Key == String, Value == String {
    func formURLEncoded() -> Data? {
        let pairs = map { key, value in
            let encodedKey = key.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? key
            let encodedValue = value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value
            return "\(encodedKey)=\(encodedValue)"
        }
        return pairs.joined(separator: "&").data(using: .utf8)
    }
}

extension Data {
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
    
    init?(base64URLEncoded string: String) {
        var base64 = string
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        
        // Add padding if needed
        let remainder = base64.count % 4
        if remainder > 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }
        
        self.init(base64Encoded: base64)
    }
}
